src/analyzer.rs

```diff
@@ -0,0 +1,212 @@
+  /// value may or may not not be for dynamic imports.
+  Known(HashMap<String, ImportAssertion>),
+}
+
+impl ImportAssertions {
+  pub fn get(&self, key: &str) -> Option<&String> {
+    match self {
+      ImportAssertions::Known(map) => match map.get(key) {
+        Some(ImportAssertion::Known(value)) => Some(value),
+        _ => None,
+      },
+      _ => None,
+    }
+  }
+}
+
+// can't use this type directly because we need to make it serialize & deserialize
+impl From<deno_ast::swc::dep_graph::ImportAssertions> for ImportAssertions {
+  fn from(value: deno_ast::swc::dep_graph::ImportAssertions) -> Self {
+    use deno_ast::swc::dep_graph::ImportAssertions::*;
+    match value {
+      None => ImportAssertions::None,
+      Unknown => ImportAssertions::Unknown,
+      Known(value) => ImportAssertions::Known(
+        value
+          .into_iter()
+          .map(|(key, value)| (key, value.into()))
+          .collect(),
+      ),
+    }
+  }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Comment {
+  pub text: String,
+  pub range: ByteRange,
+}
+
+impl Comment {
+  pub fn from_swc(
+    comment: deno_ast::swc::common::comments::Comment,
+    parsed_source: &ParsedSource,
+  ) -> Comment {
+    let range = ByteRange::from_source_range(comment.range(), parsed_source);
+    Comment {
+      text: comment.text,
+      range,
+    }
+  }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct DependencyDescriptor {
+  pub kind: DependencyKind,
+  /// A flag indicating if the import is dynamic or not.
+  pub is_dynamic: bool,
+  /// Any leading comments associated with the dependency.  This is used for
+  /// further processing of supported pragma that impact the dependency.
+  pub leading_comments: Vec<Comment>,
+  /// The range of the import/export statement.
+  pub range: ByteRange,
+  /// The text specifier associated with the import/export statement.
+  pub specifier: String,
+  /// The range of the specifier.
+  pub specifier_range: ByteRange,
+  /// Import assertions for this dependency.
+  pub import_assertions: ImportAssertions,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SpecifierWithRange {
+  pub text: String,
+  pub range: ByteRange,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum TypeScriptReference {
+  Path(SpecifierWithRange),
+  Types(SpecifierWithRange),
+}
+
+/// Provides a ModuleAnalyzer for the provided specifier.
+///
+/// It can be assumed that the source has not changed since
+/// it was loaded by deno_graph.
+pub trait ModuleAnalyzerProvider {
+  fn get_analyzer(
+    &self,
+    specifier: &ModuleSpecifier,
+    source: Arc<str>,
+    media_type: MediaType,
+  ) -> Result<Box<dyn ModuleAnalyzer>, ModuleGraphError>;
+}
+
+pub trait ModuleAnalyzer {
+  /// Converts a byte index to a `deno_graph::Position`.
+  fn byte_index_to_position(&self, byte_index: usize) -> Position;
```

Comment:
I'm going to look into removing this method and collapsing these two traits into one.